### PR TITLE
Enable sse2 for CSV parsing.

### PR DESCRIPTION
**Pull request by amosbird**
Testing data

```
select 'aaaaaaaa,bbbbbbbb,cccccccc,dddddddd,eeeeeeee,ffffffff,gggg,hhh' from numbers(3000000) into outfile '/tmp/test.csv'
```

Testing command
```
echo "select count() from file('/tmp/test.csv', CSV, 'a String, b String, c String, d String, e String, f String, g String, h String') where not ignore(e)" | clickhouse-benchmark

```

Before
```
QPS: 1.317, RPS: 3949749.687, MiB/s: 478.380, result RPS: 1.317, result MiB/s: 0.000.
0.000%  0.704 sec.
10.000% 0.712 sec.
20.000% 0.718 sec.
30.000% 0.726 sec.
40.000% 0.739 sec.
50.000% 0.754 sec.
60.000% 0.770 sec.
70.000% 0.788 sec.
80.000% 0.798 sec.
90.000% 0.815 sec.
95.000% 0.826 sec.
99.000% 0.850 sec.
99.900% 0.857 sec.
99.990% 0.858 sec.
```

After
```
QPS: 1.533, RPS: 4598308.336, MiB/s: 556.932, result RPS: 1.533, result MiB/s: 0.000.
0.000%  0.626 sec.
10.000% 0.635 sec.
20.000% 0.639 sec.
30.000% 0.642 sec.
40.000% 0.643 sec.
50.000% 0.645 sec.
60.000% 0.649 sec.
70.000% 0.652 sec.
80.000% 0.658 sec.
90.000% 0.682 sec.
95.000% 0.710 sec.
99.000% 0.727 sec.
99.900% 0.733 sec.
99.990% 0.734 sec.
```

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/src/IO/ReadHelpers.cpp [hunk 1]
```diff
@@ -552,9 +552,32 @@ void readCSVStringInto(Vector & s, ReadBuffer & buf, const FormatSettings::CSV &
         while (!buf.eof())
         {
             char * next_pos = buf.position();
-            while (next_pos < buf.buffer().end()
-                && *next_pos != delimiter && *next_pos != '\r' && *next_pos != '\n')    /// NOTE You can make a SIMD version.
-                ++next_pos;
+
+            [&]() {
```
Comment (alexey-milovidov) [inline dbms/src/IO/ReadHelpers.cpp]: Why lambda? Just code block `{...}` is Ok.

File: dbms/src/IO/ReadHelpers.cpp [hunk 2]
```diff
@@ -552,9 +552,32 @@ void readCSVStringInto(Vector & s, ReadBuffer & buf, const FormatSettings::CSV &
...
+#if __SSE2__
+                auto rc = _mm_set1_epi8('\r');
```
Comment (alexey-milovidov) [inline dbms/src/IO/ReadHelpers.cpp]: Shouldn't we include the corresponding `...intrin.h`?

File: dbms/src/IO/ReadHelpers.cpp (hunk 1 above)
Comment (amosbird) [reply dbms/src/IO/ReadHelpers.cpp]: There is a early return

File: dbms/src/IO/ReadHelpers.cpp (hunk 2 above)
Comment (amosbird) [reply dbms/src/IO/ReadHelpers.cpp]: Yeah.

**Review by alexey-milovidov — COMMENTED** [review]
That's really cool :+1:

File: dbms/src/IO/ReadHelpers.cpp [hunk 3]
```diff
@@ -552,9 +556,32 @@ void readCSVStringInto(Vector & s, ReadBuffer & buf, const FormatSettings::CSV &
...
+                auto nc = _mm_set1_epi8('\n');
+                auto dc = _mm_set1_epi8(delimiter);
+                for (; next_pos + 15 < buf.buffer().end(); next_pos += 16)
+                {
+                    __m128i bytes = _mm_loadu_si128(reinterpret_cast<const __m128i *>(next_pos));
+                    __v2du a = reinterpret_cast<__v2du>(_mm_cmpeq_epi8(bytes, rc));
+                    __v2du b = reinterpret_cast<__v2du>(_mm_cmpeq_epi8(bytes, nc));
+                    __v2du c = reinterpret_cast<__v2du>(_mm_cmpeq_epi8(bytes, dc));
+                    __m128i eq = reinterpret_cast<__m128i>(a | b | c);
```
[the remark this answers is not in the document]
Comment (amosbird) [reply dbms/src/IO/ReadHelpers.cpp]: Hmm, lemme check the output assembly. gcc folks haven't answered back yet.

[the remark this answers is not in the document]
Comment (amosbird) [reply dbms/src/IO/ReadHelpers.cpp]: They're exactly the same. I'll change to _mm_or_si128

**Pull request merged**